pkg/distributor/distributor_test.go

```diff
@@ -4449,6 +4489,104 @@ func TestSortLabels(t *testing.T) {
 	})
 }
 
+func TestDistributor_QueryStreamCorrectness(t *testing.T) {
+	const (
+		totalQueries         = 1_000
+		maxConcurrentQueries = 10
+		totalSeries          = 5
+		maxSamplesPerSeries  = 480 // 2 hours at 1 sample every 15 seconds.
+	)
+
+	orgID := "user"
+	ctx := user.InjectOrgID(context.Background(), orgID)
+
+	// Prepare matchers sets and series.
+	matchers := make([][]*labels.Matcher, totalSeries)
+	for i := 0; i < totalSeries; i++ {
+		matchers[i] = []*labels.Matcher{
+			labels.MustNewMatcher(labels.MatchEqual, labels.MetricName, fmt.Sprintf("n%d", i)),
+		}
+	}
+	series := make([]*mimirpb.PreallocTimeseries, totalSeries)
+	for i := 0; i < totalSeries; i++ {
+		ss := &mimirpb.PreallocTimeseries{
+			TimeSeries: &mimirpb.TimeSeries{
+				Labels: []mimirpb.LabelAdapter{
+					{Name: labels.MetricName, Value: fmt.Sprintf("n%d", i)},
+				},
+			},
+		}
+		totalSamples := rand.Int31n(maxSamplesPerSeries)
+		now := time.Now().Add(time.Hour * 2 * -1)
+
+		for j := int32(0); j < totalSamples; j++ {
+			ss.Samples = append(ss.Samples, mimirpb.Sample{
+				TimestampMs: now.Add(time.Second * 15 * time.Duration(j)).Unix(),
+				Value:       rand.Float64(),
+			})
+		}
+		series[i] = ss
+	}
+
+	// Set up distributors and ingesters.
+	ds, ingesters, _ := prepare(t, prepConfig{
+		numIngesters:    3,
+		happyIngesters:  3,
+		numDistributors: 1,
+	})
+	// Prepare ingester mocked series.
+	for i := 0; i < len(ingesters); i++ {
+		ingesters[i].timeseries = map[uint32]*mimirpb.PreallocTimeseries{}
+
+		for j := 0; j < totalSeries; j++ {
+			hash := shardByAllLabels(orgID, series[j].Labels)
+			ingesters[i].timeseries[hash] = series[j]
+		}
+	}
+
+	wg := sync.WaitGroup{}
+	wg.Add(totalQueries)
+
+	// Run total queries concurrently asking for all possible series.
```

Comment:
For my own understanding, what's the benefit of running the queries concurrently as part of the test?

Reply:
The `QueryStream` method invokes `queryIngesterStream` under the hood, which in turn performs the ingesters's [stream response reuse](https://github.com/grafana/mimir/blob/1f219b2c765d5b89e8a40138590b5c84ed869494/pkg/distributor/query.go#L196-L199).

So the purpose of running this concurrently is to simulate a more realistic scenario, and to validate that nothing messes up when querying ingesters.

Reply:
@ortuman thanks for clarifying! That's pretty much what I thought. Maybe you could also note this rationale in a comment (below the existing one)?

Reply:
'course! I'll include it. 👍

Reply:
done @ 74f97b01b9abb71759f5343487e411da8202c342